File: app/src/main/java/team/triplog/presentation/adapter/MainTripLogAdapter.java
package team.triplog.presentation.adapter;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;

import team.triplog.R;
import team.triplog.entity.TripLog;

public class MainTripLogAdapter extends RecyclerView.Adapter<MainTripLogAdapter.ViewHolder> {
    private Context context;
    private ArrayList<TripLog> tripLogs;
    private View.OnClickListener onClickListener;

    public MainTripLogAdapter(Context context, ArrayList<TripLog> tripLogs, View.OnClickListener onClickListener) {
        this.context = context;
        this.tripLogs = tripLogs;
        this.onClickListener = onClickListener;
    }

    @Override
    public int getItemCount() {
        return tripLogs.size();
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new ViewHolder(
                LayoutInflater
                        .from(parent.getContext())
                        .inflate(R.layout.item_main_trip_log, parent, false)
        );
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, final int position) {
        TripLog tripLog = tripLogs.get(position);
        boolean isAdd = (position == getItemCount() - 1);
        if (isAdd) {
            holder.viewGroupImage.setBackgroundResource(R.drawable.button_plus);
            holder.textGroupName.setVisibility(View.INVISIBLE);
        } else {
            holder.textGroupName.setText(tripLog.title);
//            holder.viewGroupImage.setClipToOutline(true);
//            holder.viewGroupImage.setBackgroundResource(tripLog.drawbleId);
        }

        holder.layoutMain.setTag(tripLog);
        holder.layoutMain.setOnClickListener(onClickListener);
    }

    static class ViewHolder extends RecyclerView.ViewHolder {
        TextView textGroupName;
        View viewGroupImage;
        ConstraintLayout layoutMain;

        ViewHolder(View view) {
            super(view);

            textGroupName = view.findViewById(R.id.text_group_name);
            viewGroupImage = view.findViewById(R.id.view_group_image);
            layoutMain = view.findViewById(R.id.layout_main);
        }
    }
}

File: app/src/main/java/team/triplog/presentation/main/adapter/MainTripLogAdapter.kt
package team.triplog.presentation.main.adapter

import android.annotation.SuppressLint
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import team.triplog.data.source.local.entity.TripLogEntity
import team.triplog.presentation.main.viewholder.MainTripLogViewHolder


/**
 * @author mjkim
 * @since 2021/06/06
 **/
class MainTripLogAdapter : RecyclerView.Adapter<MainTripLogViewHolder>() {

    private var items: List<TripLogEntity> = listOf()
    var listener: OnItemClickListener? = null

    override fun getItemCount(): Int = items.size

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MainTripLogViewHolder {
        return MainTripLogViewHolder.getViewHolder(parent)
    }

    override fun onBindViewHolder(holder: MainTripLogViewHolder, position: Int) {
        val item = items[position]
        val isAdd = position == itemCount - 1

        holder.bind(item, isAdd, listener)
    }

    interface OnItemClickListener {
        fun onItemClick(item: TripLogEntity)
    }

    @SuppressLint("CheckResult")
    fun setTripLogItems(tripLogItems: List<TripLogEntity>) {
        Observable.just(tripLogItems)
            .subscribeOn(AndroidSchedulers.mainThread())
            .observeOn(Schedulers.io())
            .map { DiffUtil.calculateDiff(MainTripLogDiffCallback(this.items, tripLogItems)) }
            .subscribe({
                this.items = tripLogItems
                it.dispatchUpdatesTo(this)
            }, {

            }
            )

        notifyDataSetChanged()
    }

}